### PR TITLE
Make the "icon next/prev" series load faster

### PR DESCRIPTION
**Pull request by KTibow**
## Proposed change

Right now all the buttons or icons that are next/previous choosers take 100ms to load because they use computed styles, but they can use `document.dir` to load immediately.

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

N/A. You can try playing around with the language though.

## Additional information

N/A

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

### CODE REVIEW

File: src/components/ha-icon-next.ts [hunk 1]
```diff
@@ -1,19 +1,18 @@
...
-          : mdiChevronLeft;
-    }, 100);
+  protected firstUpdated(changedProps: PropertyValues) {
+    super.firstUpdated(changedProps);
+    this.path =
+      window.getComputedStyle(this).direction === "ltr"
+        ? mdiChevronRight
+        : mdiChevronLeft;
   }
 }
```
Comment (bramkragten) [inline src/components/ha-icon-next.ts:17]: I think we can make this a lot simpler and faster by just doing:

```
@customElement("ha-icon-next")
export class HaIconNext extends HaSvgIcon {
  @property() public path =
    document.dir === "ltr" ? mdiChevronRight : mdiChevronLeft;
}
```

Comment (bramkragten) [reply src/components/ha-icon-next.ts:17]: (We should however listen for changes in direction, so either lang update or `dir` attribute, but we dont do that now either)

Comment (KTibow) [reply src/components/ha-icon-next.ts:17]: Problem is that the direction doesn't load at the start (the computed value is empty), which is why we do this in the first place.
I'm not sure how to listen to RTL changes - I did some research and couldn't find any good CSS or JS solutions, is there a way other than explicitly passing a direction to each component?

Comment (bramkragten) [reply src/components/ha-icon-next.ts:17]: > Problem is that the direction doesn't load at the start (the computed value is empty), which is why we do this in the first place.

That was when we where computing the styles, but my proposal just checks a property of the document

Comment (KTibow) [reply src/components/ha-icon-next.ts:17]: Ah, I missed that part. I'll apply that.

**Review by piitaya — APPROVED** [review]

**Comment by steverep** [comment]
```
Uncaught (in promise) Error: The following properties on element ha-icon-next will not trigger updates as expected because they are set using class fields: path. Native class fields and some compiled output will overwrite accessors used for detecting changes. See https://lit.dev/msg/class-field-shadowing for more information.
    performUpdate reactive-element.ts:1302
    scheduleUpdate reactive-element.ts:1261
    __enqueueUpdate reactive-element.ts:1233
    requestUpdate reactive-element.ts:1208
    __initialize reactive-element.ts:946
    ReactiveElement reactive-element.ts:931
    LitElement lit-element.js:105
    HaSvgIcon ha-svg-icon.ts:5
    HaIconNext ha-icon-next.ts:6
    J scoped-custom-element-registry.js:385
    c scoped-custom-element-registry.js:236
    _clone lit-html.ts:1129
    _commitTemplateResult lit-html.ts:1543
    _$setValue lit-html.ts:1384
    _update lit-html.ts:1192
    _commitTemplateResult lit-html.ts:1553
    _$setValue lit-html.ts:1384
    _commitIterable lit-html.ts:1618
    _$setValue lit-html.ts:1403
    _update lit-html.ts:1192
    _commitTemplateResult lit-html.ts:1553
    _$setValue lit-html.ts:1384
    render lit-html.ts:2183
    update lit-element.ts:165
    performUpdate reactive-element.ts:1329
    scheduleUpdate reactive-element.ts:1261
    __enqueueUpdate reactive-element.ts:1233
    requestUpdate reactive-element.ts:1208
    __initialize reactive-element.ts:946
    ReactiveElement reactive-element.ts:931
    LitElement lit-element.js:105
    SubscribeClass subscribe-mixin.ts:13
    HaConfigUpdates ha-config-updates.ts:27
    J scoped-custom-element-registry.js:385
    c scoped-custom-element-registry.js:236
    _clone lit-html.ts:1129
    _commitTemplateResult lit-html.ts:1543
    _$setValue lit-html.ts:1384
    _update lit-html.ts:1192
    _commitTemplateResult lit-html.ts:1553
    _$setValue lit-html.ts:1384
    _update lit-html.ts:1192
    _commitTemplateResult lit-html.ts:1553
    _$setValue lit-html.ts:1384
    render lit-html.ts:2183
    update lit-element.ts:165
    performUpdate reactive-element.ts:1329
    scheduleUpdate reactive-element.ts:1261
    __enqueueUpdate reactive-element.ts:1233
    requestUpdate reactive-element.ts:1208
    __initialize reactive-element.ts:946
    ReactiveElement reactive-element.ts:931
    LitElement lit-element.js:105
    SubscribeClass subscribe-mixin.ts:13
    HaConfigDashboard ha-config-dashboard.ts:119
    J scoped-custom-element-registry.js:385
3 reactive-element.ts:1302
    __enqueueUpdate reactive-element.ts:1241
    AsyncFunctionNext self-hosted:852
    (Async: async)
    requestUpdate reactive-element.ts:1208
    __initialize reactive-element.ts:946
    ReactiveElement reactive-element.ts:931
    LitElement lit-element.js:105
    HaSvgIcon ha-svg-icon.ts:5
    HaIconNext ha-icon-next.ts:6
    J scoped-custom-element-registry.js:385
    c scoped-custom-element-registry.js:236
    _clone lit-html.ts:1129
    _commitTemplateResult lit-html.ts:1543
    _$setValue lit-html.ts:1384
    _update lit-html.ts:1192
    _commitTemplateResult lit-html.ts:1553
    _$setValue lit-html.ts:1384
    _commitIterable lit-html.ts:1618
    _$setValue lit-html.ts:1403
    _update lit-html.ts:1192
    _commitTemplateResult lit-html.ts:1553
    _$setValue lit-html.ts:1384
    render lit-html.ts:2183
    update lit-element.ts:165
    performUpdate reactive-element.ts:1329
    scheduleUpdate reactive-element.ts:1261
    __enqueueUpdate reactive-element.ts:1233
    AsyncFunctionNext self-hosted:852
    (Async: async)
    requestUpdate reactive-element.ts:1208
    __initialize reactive-element.ts:946
    ReactiveElement reactive-element.ts:931
    LitElement lit-element.js:105
    SubscribeClass subscribe-mixin.ts:13
    HaConfigUpdates ha-config-updates.ts:27
    J scoped-custom-element-registry.js:385
    c scoped-custom-element-registry.js:236
    _clone lit-html.ts:1129
    _commitTemplateResult lit-html.ts:1543
    _$setValue lit-html.ts:1384
    _update lit-html.ts:1192
    _commitTemplateResult lit-html.ts:1553
    _$setValue lit-html.ts:1384
    _update lit-html.ts:1192
    _commitTemplateResult lit-html.ts:1553
    _$setValue lit-html.ts:1384
    render lit-html.ts:2183
    update lit-element.ts:165
    performUpdate reactive-element.ts:1329
    scheduleUpdate reactive-element.ts:1261
    __enqueueUpdate reactive-element.ts:1233
    AsyncFunctionNext self-hosted:852
    (Async: async)
    requestUpdate reactive-element.ts:1208
    __initialize reactive-element.ts:946
    ReactiveElement reactive-element.ts:931
    LitElement lit-element.js:105
    SubscribeClass subscribe-mixin.ts:13
    HaConfigDashboard ha-config-dashboard.ts:119
    J scoped-custom-element-registry.js:385
```

**Comment by KTibow** [comment]
@steverep while I know a way to resolve this (move setting the property into constructor), is there a correct/common way Home Assistant avoids this in general?

**Comment by steverep** [comment]
I wouldn't try to undo the decorator - just set in `willUpdate`: https://github.com/home-assistant/frontend/pull/18118

File: src/components/ha-icon-next.ts (hunk 1 above)
Comment (bramkragten) [reply src/components/ha-icon-next.ts]: You missed the `@property() public` part of my suggestion 😉

**Comment by farmio** [comment]
This change assumes that `document.dir` is always set. That seems not to be the case for HA panels that run in iFrames - resulting in wrong direction arrows for `ltr` languages there.

I'm not firm enough in FE development to say if I'm holding this wrong, or the implementation should be done more resilient. One way would be to use the already available `hass` object with [`computeRTL` from `"../common/util/compute_rtl";`](https://github.com/home-assistant/frontend/blob/0735705dd3f74e5474e20f330bb1816996ceeb2c/src/common/util/compute_rtl.ts#L4) (like in `hass-subpage`). But I have no idea if this would result in longer load times.

**Comment by KTibow** [comment]
> One way would be to use the already available hass object with [computeRTL from "../common/util/compute_rtl";](https://github.com/home-assistant/frontend/blob/0735705dd3f74e5474e20f330bb1816996ceeb2c/src/common/util/compute_rtl.ts#L4) (like in hass-subpage).

These icons are used in environments like the login page so that wouldn't work. Consider creating a PR that either falls back to ltr if unset, or sets `document.dir` using the `computeRTL` function.

**Pull request merged**